app.go

```diff
@@ -296,9 +296,12 @@ type App struct {
 	errorHooks []ErrorHandler
 	validate   bool
 	// Used to signal shutdowns.
-	donesMu     sync.Mutex // guards dones and shutdownSig
-	dones       []chan os.Signal
-	shutdownSig os.Signal
+	donesMu        sync.Mutex // guards dones and shutdownSig
+	dones          []chan os.Signal
+	shutdownSig    os.Signal
+	waitsMu        sync.Mutex // guards waits and shutdownCode
+	waits          []chan ShutdownSignal
+	shutdownSignal *ShutdownSignal
```

Comment:
I think we can simplify state management here a bit by collapsing dones and waits into a single list/mutex and recording only shutdownSignal, not shutdownSig.

Basically, let's implement Done (the weaker API) in terms of Wait (the richer API).
The problem is `Wait()` returns a `chan ShutdownSignal` and `Done()` returns a `chan os.Signal`.
Mapping ShutdownSignal to Signal is possible, but using the current setup, requires a goroutine or similar to map. That's not a good idea.

Instead, we can unify the internal state here with an interface for the signal receiver (the channel that takes os.Signal or ShutdownSignal from App) and "discard" the extra information from ShutdownSignal in the os.Signal implementation.

Basically, imagine a private interface:

```go
// Returns an error if the channel wasn't receiving.
type signalRecevier interface{ ReceiveSignal(ShutdownSignal) error }
```

Then the two implementations are,

```go
type osSignalReceiver struct{ ch chan os.Signal<- }

func (r *osSignalReceiver) ReceiveSignal(sig ShutdownSignal) error {
  select {
  case r.ch <- sig.Signal:
    return nil
  default:
    return errChannelBlocked  // define this somewhere
  }
}
```

The shutdownSignalReceiver implementation will do the same but it'll stick the `sig` in as-is.

Then internal state management is just a `[]signalReceiver` instead of `[]chan os.Signal` and `[]chan ShutdownSignal`, and the os.Signal field can be dropped completely from App.

Both, App.Wait and App.Done will just build a signalReceiver and call a private method.
Something like this setup.

```go
func (app *App) Wait() <-chan ShutdownSignal {
  c := make(chan ShutdownSignal, 1)
  app.onShutdown(&shutdownSignalReceiver{ch: c})
  return c
}

func (app *App) Done() <-chan os.Signal {
  c := make(chan os.Signal, 1)
  app.onShutdown(&osSignalReceiver{ch: c})
  return c
}
```